src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,46 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> lastNode;
+    private int size;
 
+    public void push(T value) {
+        lastNode = new Node<>(lastNode, value);
+        size++;
     }
 
     public T peek() {
-        return null;
+        sizeCheck();
+        return lastNode.value;
     }
 
     public T pop() {
-        return null;
+        sizeCheck();
+        Node<T> oldNode = lastNode;
+        lastNode = lastNode.previous;
+        size--;
+        return oldNode.value;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private void sizeCheck() {
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> previous;
+
+        public Node(Node<T> previous, T value) {
```

Comment:
```suggestion
        private Node(Node<T> previous, T value) {
```
will the constructor be available from outside if the Node class is private?